src/Rank.cpp

```diff
@@ -1100,6 +1100,29 @@ void Rank::addTremulantToPipes(
 
 		// if there are any matching attacks we add them
 		if (!pipeAttacksToAdd.IsEmpty()) {
+			// warn if all previous attacks ignore wave tremulant
```

Comment:
If multiple attacks exist, then only a warning is shown.  If the bool `setSingle` is `false`, then no attack will be modified.  No method is provided to toggle `setSingle`, a user accessible check box will need to be added.

Another issue I was trying to solve was when there are less tremulant pipes than non-tremulant pipes.  This is the case in the Barton SoloStrings and SoloStringsTrem ranks I had.  I was hoping to avoid error prone hand editing to ignores-tremulant all the non-tremulant pipes without matching tremulant pipes.

I think that having at least a warning would be a good idea:  "Have you forgotten to select Load pipes to play when wave tremulant is off"
1114 becomes` if (!hasNonWave) {`
remove lines 1106, 1117-1125

Reply:
> If multiple attacks exist, then only a warning is shown.

But how to know if it's intentional or not.

> I think that having at least a warning would be a good idea

Yes, that was what I suggested in my initial comment. One should parse **all** the attacks, and if not at least one of them has "play when wave tremulant is off" - then issue a warning (when loading the wave tremulant samples). Still this might be perfectly valid since the user can load the non-tremmed versions later. I outlined the recommended procedure above.

Reply:
Pushed the wrong button and lost my much more detailed response!  Oh well.  Older, wiser (jury's out on that!).

Fortunately GoOdf does not need to actually read the samples.  I have a script that generates a test sample-set tree (empty *.wav).  I'll look at coding a warning popup.

Reply:
As I wrote in the README.md file:

" Always work on copies or on files you don't mind having destroyed!"

Anyway, thanks for trying to improve the program!

Reply:
I have a mental block.  Do you have an example of a real-world ODF that uses all three states of `isTremulant` (NULL, true, false) on a single pipe in a meaningful way?  A rank with a mix of invdividual pipes with `isTremulant` NULL, **or** `isTremulant` defined makes perfect sense.  All three states in a single pipe, not so much.

Here's a contrived single pipe with four attacks. Attacks 0 and 1 will always be selected,  with attack 2 when Tremulant is on, and attack 3 when Tremulant is off.

```
Pipe001=hwfake8\036.wav
Pipe001AttackCount=3
Pipe001Attack001=hwfake8\036_bis.wav
Pipe001Attack002=hwfake8_tr\036.wav
Pipe001Attack002IsTremulant=1
Pipe001Attack003=hwfake8_rel\036.wav
Pipe001Attack003IsTremulant=0
```

| Attack | Switch | IsTremulant | Played |
|--------|--------|--------|--------|
| 0 | ON | NULL | Yes |
| 1 | ON | NULL | Yes |
| 2 | ON | 1 | Yes |
| 3 | ON | 0 | No |
| 0 | OFF | NULL | Yes |
| 1 | OFF | NULL | Yes |
| 2 | OFF | 1 | No |
| 3 | OFF | 0 | Yes |

Reply:
I've pushed 382de96.  No attacks/releases are modified, and only one Dialog is displayed per rank update if there are any detected  isTremulant mixes.  More places where this could happen were uncovered in testing, so there are more lines changed than expected!

Reply:
@larspalo would you be able to run https://gist.github.com/kwhite-uottawa/4c1486e06a873c23d250d28228b0cf52#file-checktrem-sh on your .organ files?  It should be silent unless there are pipes with three tremulant states.  It's a script that runs awk to convert Pipe* lines to SQL which is then to be piped to sqlite3.

Output over the .organ files I have collected `find ../../Organs/ -name '*.organ' | sed -e 's+.*+./checktrem.sh "&" | sqlite3+' | sh -v 2>&1 | tee checktrem.log` is in https://gist.github.com/kwhite-uottawa/4c1486e06a873c23d250d28228b0cf52#file-checktrem-log

Output for suspect .organ files looks like this:
```
./checktrem.sh "../../Organs/Buckerburg/Audio/BuckerburgODF-trem.organ" | sqlite3
rank     name               pipeno  what     nsamples  nnulls  ndefs  toterrs
-------  -----------------  ------  -------  --------  ------  -----  -------
Stop016  37 Quintadena 16'  001     attack   2         1       1      300    
Stop016  37 Quintadena 16'  001     release  5         3       2      300    
Stop016  37 Quintadena 16'  002     attack   2         1       1      300    
Stop016  37 Quintadena 16'  002     release  5         3       2      300    
Stop016  37 Quintadena 16'  003     attack   2         1       1      300     
```

Reply:
I've got no problem reported for most sample sets, nor did I expect it. Barton is a special case of course. Personally, I don't see much point in fussing about the tremulant states either. A simple warning, well phrased, if the case is detected is absolutely good enough for me.

Reply:
> I've got no problem reported for most sample sets, nor did I expect it. Barton is a special case of course. Personally, I don't see much point in fussing about the tremulant states either. A simple warning, well phrased, if the case is detected is absolutely good enough for me.

Barton is indeed a special case!  Doing the right thing for a beginning GoOdf user with that sample set can be a challenge.

I had removed updating the tremulant and only display a warning message in 382de9684d6ee6d44132c0153d8a34d84146e282 Better phrasing would be helpful.  I just don't have any ideas.
```
wxMessageDialog msg(::wxGetApp().m_frame, wxT("Warning: A mix of tremulant and non-tremulant pipes in a rank."), wxT("Pipe Tremulants Warning"), wxOK|wxCENTRE|wxICON_EXCLAMATION);
```